plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs

```diff
@@ -1128,69 +1185,118 @@ instance uni ~ DefaultUni => ToBuiltinMeaning uni DefaultFun where
             {-# INLINE costingFun #-}
         -- See Note [Builtin semantics variants]
         in case semvar of
-            DefaultFunSemanticsVariant1 -> makeBuiltinMeaning
-               @(Integer -> BS.ByteString -> BS.ByteString)
-               (\n xs -> BS.cons (fromIntegral @Integer n) xs)
-               costingFun
+            DefaultFunSemanticsVariant1 ->
+                let consByteStringDenotation :: Integer -> BS.ByteString -> BS.ByteString
+                    consByteStringDenotation n xs = BS.cons (fromIntegral n) xs
+                    {-# INLINE consByteStringDenotation #-}
+                in makeBuiltinMeaning
+                    consByteStringDenotation
+                    costingFun
             -- For builtin semantics variants other (i.e. larger) than
             -- DefaultFunSemanticsVariant1, the first input must be in range
             -- [0..255].  See Note [How to add a built-in function: simple
             -- cases]
-            DefaultFunSemanticsVariant2 -> makeBuiltinMeaning
-              @(Word8 -> BS.ByteString -> BS.ByteString)
-              BS.cons
-              costingFun
+            DefaultFunSemanticsVariant2 ->
+                let consByteStringDenotation :: Word8 -> BS.ByteString -> BS.ByteString
```

Comment:
> At PLC level this one takes an `integer` as the first argument and fails if it's not in the range 0..255 (the other variant reduces it modulo 256 and never fails). 

Correct.

> Is this the right type for that? We don't need `EvaluationResult` here because the failure occurs during the conversion from `Integer` to `Word8`

Yes, exactly.

> before we even get to the denotation, is that right?

Yes, before we get to the denotation (i.e. `consByteStringDenotation`), but within the runtime denotation (i.e. what `consByteStringDenotation` becomes after `makeBuiltinMeaning` adds unlifting/lifting and costing to it).

Note that even if there was no `Word8`, there would still be possibility of evaluation failure, since the user may supply a `Bool` or even a lambda for the `ByteString` argument.

Reply:
We could use fromInteger instead of fromIntegral in the semvar==1 case, but I guess ghc inlines fromIntegral , so it should be the same in the ghc core

Reply:
Just checked out of curiosity, it inlines to `integerToWord#` and later `wordToWord8#` is called over the result, so we probably can't do better than that.